### PR TITLE
fix: propagate retracts before updates in order to have clean indexes

### PR DESCRIPTION
**Pull request by triceo**
If an update of a tuple is sent into a join node before a retract of that same tuple,
the joiner is effectively using stale information and this can lead to various issues
such as NPEs. (See the newly introduced BavetRegressionTest.)

To address this, we make three changes to how tuples are propagated:

First, the dirty queues (now called propagation queues) order their tuples and make sure that retracts are propagated before updates, and updates before inserts.
StaticPropagationQueue handles most nodes, 
but DynamicPropagationQueue is necessary for GroupNode and IfExistsNode, which need more sophisticated (and slower) reordering behavior.

Second, nodes are distributed into layers, each layer has nodes that depend on nodes from previous layers. 
Propagations happen per layer; all nodes in a layer propagate, and then we move on to the next layer. 
Within each layer, all nodes first propagate retracts, then all nodes propagate updates, then all nodes propagate inserts.

Finally, specialized ForEach nodes are introduced for cases with and without nullity filters.
When a nullity filter is used, the ForEach node is now smart enough to trigger an insert only when a predicate matches, 
and therefore subsequent updates/retracts may avoid propagation altogether.

Together, these changes significantly clean up and standardize tuple propagation.
Performance drop of at most 15 % was seen.
Some cases benefit by as much as 5 %, perhaps due to the elimination of useless join cross-products
that no longer result from join updates over tuples that would be later retracted.

The average performance drop over all the measured use cases is around 5 %.
Considering that these changes address an incorrect behavior, 
the performance impact has to be accepted.

All turtle tests pass locally.

### CODE REVIEW

File: core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/bi/BavetAbstractBiConstraintStream.java [hunk 1]
```diff
@@ -159,9 +159,10 @@ private <C> BiConstraintStream<A, B> ifExistsOrNot(boolean shouldExist, UniConst
     @Override
     public <ResultContainer_, Result_> UniConstraintStream<Result_> groupBy(
             BiConstraintCollector<A, B, ResultContainer_, Result_> collector) {
-        GroupNodeConstructor<UniTuple<Result_>> nodeConstructor = of((groupStoreIndex, undoStoreIndex, tupleLifecycle,
-                outputStoreSize, environmentMode) -> new Group0Mapping1CollectorBiNode<>(groupStoreIndex, undoStoreIndex,
-                        collector, tupleLifecycle, outputStoreSize, environmentMode));
+        GroupNodeConstructor<UniTuple<Result_>> nodeConstructor =
```
Comment (triceo) [inline core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/bi/BavetAbstractBiConstraintStream.java:162]: Note for reviewers: there is *nothing interesting* in these `*ConstraintStream` changes. Only mechanical updates, pure boilerplate.

File: core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/bi/Group0Mapping1CollectorBiNode.java [hunk 2]
```diff
@@ -13,7 +13,8 @@ final class Group0Mapping1CollectorBiNode<OldA, OldB, A, ResultContainer_>
     public Group0Mapping1CollectorBiNode(int groupStoreIndex, int undoStoreIndex,
             BiConstraintCollector<OldA, OldB, ResultContainer_, A> collector,
             TupleLifecycle<UniTuple<A>> nextNodesTupleLifecycle, int outputStoreSize, EnvironmentMode environmentMode) {
-        super(groupStoreIndex, undoStoreIndex, null, collector, nextNodesTupleLifecycle, environmentMode);
+        super(groupStoreIndex, undoStoreIndex,
```
Comment (triceo) [inline core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/bi/Group0Mapping1CollectorBiNode.java:16]: Note for reviewers: there is *nothing interesting* in these `Group*` changes. Only mechanical updates, pure boilerplate.

All the interesting stuff is typically in `Abstract*` classes.

**Review by rsynek — APPROVED** [review]

File: core/core-impl/src/main/java/ai/timefold/solver/core/impl/util/ElementAwareList.java [hunk 3]
```diff
@@ -96,7 +96,22 @@ public T next() {
...
+        switch (size) {
+            case 0 -> {
+                return "[]";
+            }
+            case 1 -> {
+                return "[" + first.getElement() + "]";
+            }
+            default -> {
+                StringBuilder builder = new StringBuilder("[");
+                for (T entry : this) {
```
[the remark this answers is not in the document]
Comment (triceo) [reply core/core-impl/src/main/java/ai/timefold/solver/core/impl/util/ElementAwareList.java:108]: Typical, hard to say - could be tens, hundreds, thousands.
Maximal - there is no upper limit; as many as the indexes need.

File: core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/common/AbstractDynamicPropagationQueue.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+import ai.timefold.solver.constraint.streams.bavet.common.tuple.TupleState;
+
+/**
+ * This implementation has the capability to move tuples between the individual propagation queues.
+ * This is significantly less efficient than the {@link StaticPropagationQueue}.
+ *
+ * @param <Carrier_>
+ * @param <Tuple_>
+ */
+sealed abstract class AbstractDynamicPropagationQueue<Carrier_ extends AbstractPropagationMetadataCarrier, Tuple_ extends AbstractTuple>
```
[the remark this answers is not in the document]
Comment (triceo) [reply core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/common/AbstractDynamicPropagationQueue.java:19]: Hmm. Yes, but... isn't the literally thousands of other tests enough to test this?

[the remark this answers is not in the document]
Comment (rsynek) [reply core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/common/AbstractDynamicPropagationQueue.java:19]: Hard to say without the Sonar PR analysis. Let's see the overall analysis after merging.

File: core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/common/AbstractGroupNode.java [hunk 5]
```diff
@@ -181,82 +184,74 @@ public void update(InTuple_ tuple) {
...
-                case DEAD:
-                default:
-                    throw new IllegalStateException("Impossible state: The group (" + oldGroup + ") in node (" +
-                            this + ") is in an unexpected state (" + outTuple.getState() + ").");
+            var outTuple = accumulate(tuple, oldGroup);
+            switch (outTuple.state) {
+                case CREATING, UPDATING -> {
+                    // Already in the correct state.
+                }
+                case OK -> propagationQueue.update(oldGroup);
```
Comment (rsynek) [inline core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/common/AbstractGroupNode.java:196]: Don't we need to set the tuple state to UPDATING anymore?

**Review by rsynek — APPROVED** [review]
Looks good. I left there a couple of questions where I was not sure if that code change was intended, please have a look.

File: core/constraint-streams/src/test/java/ai/timefold/solver/constraint/streams/bavet/BavetRegressionTest.java [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+
+        TestdataSolution solution = TestdataSolution.generateSolution(1, 2);
+        TestdataEntity entity1 = solution.getEntityList().get(0);
+        TestdataEntity entity2 = solution.getEntityList().get(1);
+        TestdataValue value = solution.getValueList().get(0);
+        entity1.setValue(null);
+        entity2.setValue(value);
+
+        scoreDirector.setWorkingSolution(solution);
+        // TODO update the assertions once no longer throwing exceptions
```
[the remark this answers is not in the document]
Comment (triceo) [reply core/constraint-streams/src/test/java/ai/timefold/solver/constraint/streams/bavet/BavetRegressionTest.java:215]: Good catch.

**Pull request merged**